include/crab/domains/symbolic_variable_eq_domain.hpp

```diff
@@ -0,0 +1,1031 @@
+  /// TODO: to improve sharable results, we need to change data structure.
+  this_domain_t join(const this_domain_t &left,
+                         const this_domain_t &right) const {
+    this_domain_t res;
+    var_id current_id = 0;
+    // res.dump();
+    // for (auto &kv : left.m_parents) { // for each k , v
+    for (auto it = left.m_parents.begin(); it != left.m_parents.end(); it++) {
+      const element_t &k = it->first;
+      const element_t &v = it->second;
+      if (k == v) {
+        continue;
+      }
+      auto it_k = right.m_parents.find(k);
+      auto it_v = right.m_parents.find(v);
+      // check if k == v exists in another map
+      if (it_k != right.m_parents.end() && it_v != right.m_parents.end() &&
+          it_k->second == it_v->second) {
+        res.m_parents.insert({v, k}); // insert new pair <v, k>
+        if (!res.contains(k)) {
+          std::shared_ptr<domain_t> absval_ptr =
+              std::make_shared<domain_t>(current_id);
+          res.make_set(k, absval_ptr);
+          current_id++;
+        }
+      }
+      // for (auto &kv2 : left.m_parents) { // for each k2 , v2
+      for (auto it2 = std::next(it); it2 != left.m_parents.end(); it2++) {
+        const element_t &k2 = it2->first;
+        const element_t &v2 = it2->second;
+        if (v != v2 || (k == k2 && v == v2)) {
+          // not in k's class or
+          // k2, v2 is the same as k, v
+          continue;
+        }
+        // check if k2 == k exists in another map
+        auto it_k2 = right.m_parents.find(k2);
+        if (it_k != right.m_parents.end() && it_k2 != right.m_parents.end() &&
+            it_k2->second == it_k->second) {
+          res.m_parents.insert({k2, k}); // insert new pair <k2, k>
+          if (!res.contains(k)) {
+            std::shared_ptr<domain_t> absval_ptr =
+                std::make_shared<domain_t>(current_id);
+            res.make_set(k, absval_ptr);
+            current_id++;
+          }
+        }
+      }
+    }
+    res.normalize();
+    return res;
+  }
+
+  /// @brief a helper method for domain operation meet
+  /// @param left one abstract state
+  /// @param right one abstract state
+  /// @return a new abstract state saved the meet result
+  /// TODO: This operation is not an optimal solution. This computation does
+  /// not follow the feature of union-find data structure.
+  /// The solution is merging two equivalence classes if two equivalence classes
+  /// shared common elements. As the result, it computed a set of a set of
+  /// equivalence classes where each class is a super set of two or more classes
+  /// from the given two abstract states. That is, the result of the meet will
+  /// keep equalities appeared either in left or in right states.
+  /// Formally, forall cls_x \in left. forall cls_y \in right ::
+  ///           cls_x ^ cls_y != \empty => cls_x union cls_y
+  this_domain_t meet(const this_domain_t &left,
```

Comment:
Building the classes from left and right is not efficient. Try doing this:

Initialize `res` as a copy of `left` (or `right`). For every equality `a = b` in `right` (or `left`), call `res.union(a,b)`.

What about abstract values of the classes, are you doing the meet for those?